sql/src/main/java/io/crate/memory/MemoryManager.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.memory;
+
+import io.netty.buffer.ByteBuf;
+
+/**
+ * Instances of this interface can be used to allocate `ByteBuf` instances.
+ *
+ * <ul>
+ *     <li>Components that create a MemoryManager are responsible for closing it</li>
+ *     <li>Components that use a MemoryManager to allocate ByteBuf instances MUST NOT release them.
+ *     The MemoryManager is responsible for releasing the ByteBuf instances it creates</li>
+ *  </ul>
+ */
+public interface MemoryManager extends AutoCloseable {
+
+    /**
+     * TODO: ByteBuf instances could be re-sized and we wouldn't be able to track this.
+     *      Maybe we have to go with another structure that allows re-size operation tracking.
```

Comment:
Maybe we could return a ByteBuf wrapper which accounts the new capacity on wrapping the `capacity(int newCapacity`) method?

Reply:
+1 for a ByteBuf wrapper and observing `ByteBuf#capacity`

Reply:
Do you think it is fine to leave it as is for now? The current use-case (HLL) doesn't resize. 

We can follow up on this once it becomes a problem.